osu.Game/Overlays/BeatmapSet/Buttons/FavouriteButton.cs

```diff
@@ -55,28 +42,49 @@ private void load()
                     Size = new Vector2(18),
                     Shadow = false,
                 },
+                loading = new DimmedLoadingLayer(0.8f, 0.5f),
             });
 
+            favourited.ValueChanged += favourited => icon.Icon = favourited.NewValue ? FontAwesome.Solid.Heart : FontAwesome.Regular.Heart;
+
             BeatmapSet.BindValueChanged(setInfo =>
             {
-                if (setInfo.NewValue?.OnlineInfo?.HasFavourited == null)
+                Enabled.Value = BeatmapSet.Value?.OnlineBeatmapSetID > 0;
+                favourited.Value = setInfo.NewValue?.OnlineInfo?.HasFavourited ?? false;
+            }, true);
+
+            Action = () =>
+            {
+                if (loading.State.Value == Visibility.Visible)
                     return;
 
-                favourited.Value = setInfo.NewValue.OnlineInfo.HasFavourited;
-            });
+                // guaranteed by disabled state abvove.
+                Debug.Assert(BeatmapSet.Value.OnlineBeatmapSetID != null);
 
-            favourited.ValueChanged += favourited =>
-            {
-                if (favourited.NewValue)
+                loading.Show();
+
+                request?.Cancel();
+
+                request = new PostBeatmapFavouriteRequest(BeatmapSet.Value.OnlineBeatmapSetID.Value, favourited.Value ? BeatmapFavouriteAction.UnFavourite : BeatmapFavouriteAction.Favourite);
+
+                request.Success += () =>
                 {
-                    pink.FadeIn(200);
-                    icon.Icon = FontAwesome.Solid.Heart;
-                }
-                else
+                    favourited.Toggle();
+                    loading.Hide();
+                };
+
+                request.Failure += e =>
                 {
-                    pink.FadeOut(200);
-                    icon.Icon = FontAwesome.Regular.Heart;
-                }
+                    notifications.Post(new SimpleNotification
```

Comment:
This leads to `TestSceneBeatmapSetOverlay.TestAvailability()` crashing the test browser - clicking the button in that test raises a 404, which then causes an attempt to access `notifications` which is null inside the test. This was a problem before too. Unsure if we just want to null-check this, refactor somehow or leave as-is.

Reply:
should null check as null notifications overlay is allowed here.